File: admin/src/services/email.ts
import { request } from './http'


export interface EmailTemplate {
    id: number
    code: string
    name: string
    eventName: string
    subjectTemplate: string
    htmlTemplate: string
    textTemplate: string
    toEmails: string[]
    isEnabled: boolean
    createdAt: string
    updatedAt: string
}

export interface CreateEmailTemplateReq {
    code: string
    name: string
    eventName: string
    subjectTemplate: string
    htmlTemplate: string
    textTemplate: string
    toEmails: string[]
    isEnabled: boolean
}

export interface UpdateEmailTemplateReq {
    name: string
    eventName: string
    subjectTemplate: string
    htmlTemplate: string
    textTemplate: string
    toEmails: string[]
    isEnabled: boolean
}

export interface EmailTemplatePreviewReq {
    variables: Record<string, any>
}

export interface EmailTemplatePreviewResp {
    subject: string
    htmlBody: string
    textBody: string
}

export interface EmailTemplateTestReq {
    toEmails: string[]
    variables: Record<string, any>
}

export interface EmailSubscription {
    id: number
    email: string
    eventName: string
    status: 'pending' | 'verified' | 'unsubscribed'
    sourceIp: string
    unsubscribedAt?: string
    token?: string
    createdAt: string
    updatedAt: string
}

export interface EmailSubscriptionListResp {
    items: EmailSubscription[]
    total: number
    page: number
    size: number
}

export interface EmailSubscriptionListParams {
    page?: number
    pageSize?: number
    eventName?: string
    status?: string
    search?: string
}

export interface BatchUpdateEmailSubscriptionStatusReq {
    ids: number[]
    status: string
}

function stripEmpty<T extends Record<string, any>>(value: T) {
    return Object.fromEntries(
        Object.entries(value).filter(([, entry]) => entry !== undefined && entry !== null && entry !== ''),
    ) as T
}

// 邮件事件相关 API

// 邮件模板相关 API
export function listEmailTemplates() {
    return request<EmailTemplate[]>('/admin/email/templates', {
        method: 'GET',
    })
}

export function createEmailTemplate(data: CreateEmailTemplateReq) {
    return request<EmailTemplate>('/admin/email/templates', {
        method: 'POST',
        body: data,
    })
}

export function updateEmailTemplate(code: string, data: UpdateEmailTemplateReq) {
    return request<EmailTemplate>(`/admin/email/templates/${code}`, {
        method: 'PUT',
        body: data,
    })
}

export function deleteEmailTemplate(code: string) {
    return request<void>(`/admin/email/templates/${code}`, {
        method: 'DELETE',
    })
}

export function previewEmailTemplate(code: string, data: EmailTemplatePreviewReq) {
    return request<EmailTemplatePreviewResp>(`/admin/email/templates/${code}/preview`, {
        method: 'POST',
        body: data,
    })
}

export function testEmailTemplate(code: string, data: EmailTemplateTestReq) {
    return request<void>(`/admin/email/templates/${code}/test`, {
        method: 'POST',
        body: data,
    })
}

export function getTemplateByCode(code: string) {
    return request<EmailTemplate>(`/admin/email/templates/${code}`, {
        method: 'GET',
    })
}

// 邮件订阅相关 API
export function listEmailSubscriptions(params: EmailSubscriptionListParams) {
    return request<EmailSubscriptionListResp>('/admin/email/subscriptions', {
        method: 'GET',
        query: stripEmpty(params),
    })
}

export function batchUpdateEmailSubscriptionStatus(data: BatchUpdateEmailSubscriptionStatusReq) {
    return request<void>('/admin/email/subscriptions/status', {
        method: 'PUT',
        body: data,
    })
}

File: server/internal/infra/persistence/model/email.go
package model

import (
	"time"

	"gorm.io/datatypes"
	"gorm.io/gorm"
)

type EmailTemplate struct {
	ID              int64          `gorm:"column:id;primaryKey"`
	Code            string         `gorm:"column:code;size:80;not null;uniqueIndex:uq_email_template_code"`
	Name            string         `gorm:"column:name;size:120;not null"`
	EventName       string         `gorm:"column:event_name;size:120;not null"`
	SubjectTemplate string         `gorm:"column:subject_template;type:text;not null"`
	HTMLTemplate    string         `gorm:"column:html_template;type:text;not null"`
	TextTemplate    string         `gorm:"column:text_template;type:text;not null"`
	ToEmails        datatypes.JSON `gorm:"column:to_emails;type:jsonb;not null"`
	IsEnabled       bool           `gorm:"column:is_enabled;not null"`
	CreatedAt       time.Time      `gorm:"column:created_at;autoCreateTime"`
	UpdatedAt       time.Time      `gorm:"column:updated_at;autoUpdateTime"`
	DeletedAt       gorm.DeletedAt `gorm:"column:deleted_at;index"`
}

func (EmailTemplate) TableName() string { return "email_template" }

type EmailOutbox struct {
	ID           int64          `gorm:"column:id;primaryKey"`
	TemplateID   *int64         `gorm:"column:template_id"`
	TemplateCode string         `gorm:"column:template_code;size:80;not null"`
	EventName    string         `gorm:"column:event_name;size:120;not null"`
	ToEmails     datatypes.JSON `gorm:"column:to_emails;type:jsonb;not null"`
	Subject      string         `gorm:"column:subject;type:text;not null"`
	HTMLBody     string         `gorm:"column:html_body;type:text;not null"`
	TextBody     string         `gorm:"column:text_body;type:text;not null"`
	Status       string         `gorm:"column:status;size:20;not null"`
	RetryCount   int            `gorm:"column:retry_count;not null"`
	NextRetryAt  time.Time      `gorm:"column:next_retry_at;not null"`
	LastError    string         `gorm:"column:last_error;type:text"`
	SentAt       *time.Time     `gorm:"column:sent_at"`
	CreatedAt    time.Time      `gorm:"column:created_at;autoCreateTime"`
	UpdatedAt    time.Time      `gorm:"column:updated_at;autoUpdateTime"`
}

func (EmailOutbox) TableName() string { return "email_outbox" }

type EmailSubscription struct {
	ID             int64      `gorm:"column:id;primaryKey"`
	Email          string     `gorm:"column:email;size:255;not null"`
	EventName      string     `gorm:"column:event_name;size:120;not null"`
	Status         string     `gorm:"column:status;size:20;not null"`
	Token          string     `gorm:"column:token;size:80;not null;uniqueIndex:uq_email_subscription_token"`
	SourceIP       string     `gorm:"column:source_ip;size:45"`
	UnsubscribedAt *time.Time `gorm:"column:unsubscribed_at"`
	CreatedAt      time.Time  `gorm:"column:created_at;autoCreateTime"`
	UpdatedAt      time.Time  `gorm:"column:updated_at;autoUpdateTime"`
}

func (EmailSubscription) TableName() string { return "email_subscription" }
